### PR TITLE
VB-3915 Store booker reference, new action_by table and migration

### PR DESCRIPTION
**Pull request by AledWEvans**
Adds new way of storing the booker reference.

### CODE REVIEW

File: src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        user_type               VARCHAR(80)     NOT NULL,
+        create_timestamp        timestamp       NOT NULL default current_timestamp
+    );
+
+    CREATE INDEX idx_action_by_user_type ON actioned_by(user_type);
+
+    INSERT INTO actioned_by(user_name,user_type)
+            SELECT actioned_by,user_type FROM event_audit
+                                         WHERE "type" not in ('PRISON_VISITS_BLOCKED_FOR_DATE','PRISONER_RELEASED_EVENT')
+                                         group by actioned_by,user_type;
```
Comment (AledWEvans) [inline src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql:15]: Should I bother doing PUBLIC userType here also?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt [hunk 2]
```diff
@@ -22,12 +23,11 @@ interface EventAuditRepository : JpaRepository<EventAudit, Long> {
   fun findLastBookedVisitEventByBookingReference(bookingReference: String): EventAudit?
 
   @Query(
-    "SELECT actioned_by FROM event_audit " +
-      "WHERE booking_reference = :bookingReference AND type in ('UPDATED_VISIT','BOOKED_VISIT') " +
-      "ORDER BY id DESC LIMIT 1 ",
-    nativeQuery = true,
+    "SELECT ea.actionedBy FROM EventAudit ea " +
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt:26]: Question: why did you change this from native SQL to JPQL?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitEventAuditService.kt [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+  @Lazy
+  @Autowired
+  private lateinit var actionedByRepository: ActionedByRepository
+
+  @Lazy
+  @Autowired
+  private lateinit var eventAuditRepository: EventAuditRepository
+
+  fun saveEventAudit(
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitEventAuditService.kt:55]: Suggestion: Is it worth giving these methods their own names? As they technically save "different" eventAudits?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitEventAuditService.kt [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+  @Transactional(readOnly = true)
+  fun getLastEventForBooking(bookingReference: String): EventAuditDto? {
+    return eventAuditRepository.findLastBookedVisitEventByBookingReference(bookingReference)?.let {
+      EventAuditDto(it)
+    }
+  }
+
+  @Transactional(readOnly = true)
+  fun getLastUserToUpdateToSlotByReference(bookingReference: String): String {
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitEventAuditService.kt:175]: rename to: getLastUserToUpdateSlotByReference ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt [hunk 5]
```diff
@@ -1,58 +1,74 @@
...
 
-  @Column(nullable = false)
-  var actionedBy: String,
+  @Column(name = "ACTIONED_BY_ID", nullable = true)
+  private val actionedById: Long? = null,
 
-  @Enumerated(EnumType.STRING)
-  @Column(name = "user_type", nullable = false)
-  val userType: UserType,
+  @ManyToOne(fetch = FetchType.LAZY, cascade = [REFRESH, PERSIST, MERGE])
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt:53]: probably don't,... refresh makes sense...

File: src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+
+    -- For Notification as they do not have a source
+    INSERT INTO actioned_by(user_type) VALUES ('SYSTEM');
```
Comment (adaviesMOJ) [inline src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql:19]: Question: How does this insert statement work? There is no select, it just seems to be inserting directly with no condition?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt:26]: it was an easy way to return the actionedBy with out having to do a join

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitEventAuditService.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitEventAuditService.kt:55]: I did consider this was in two minds,... so decided to go with the overload approach.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/BookVisitTest.kt [hunk 7]
```diff
@@ -647,10 +682,18 @@ class BookVisitTest : IntegrationTestBase() {
...
 
-  private fun assertAuditEvent(visitDto: VisitDto, visitEntity: Visit) {
+  private fun assertAuditEvent(visitDto: VisitDto, visitEntity: Visit, userType: UserType = STAFF) {
     val eventAudit = this.eventAuditRepository.findLastEventByBookingReference(visitDto.reference)
     assertThat(eventAudit.type).isEqualTo(EventAuditType.BOOKED_VISIT)
-    assertThat(eventAudit.actionedBy).isEqualTo("booking_guy")
+    assertThat(eventAudit.actionedBy).isNotNull()
+    assertThat(eventAudit.actionedBy.userType).isEqualTo(userType)
+    if (STAFF == eventAudit.actionedBy.userType) {
+      assertThat(eventAudit.actionedBy.userName).isEqualTo("booking_guy")
```
Comment (adaviesMOJ) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/BookVisitTest.kt:691]: could assert that the other field is null ? To make sure we're only inserting into the 1 column per user type?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/BookVisitTest.kt [hunk 8]
```diff
@@ -647,10 +682,18 @@ class BookVisitTest : IntegrationTestBase() {
...
+    }
+    if (PUBLIC == eventAudit.actionedBy.userType) {
+      assertThat(eventAudit.actionedBy.bookerReference).isEqualTo("booking_guy")
```
Comment (adaviesMOJ) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/BookVisitTest.kt:694]: same here

could assert that the other field is null ? To make sure we're only inserting into the 1 column per user type?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt:53]: what does refresh do? How would that column be "refreshed" ?

File: src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql (hunk 6 above)
Comment (AledWEvans) [reply src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql:19]: Thats exactly what its doing, it connects the data later on in line 75

**Review by adaviesMOJ — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt]: I means you can refresh the object from the DB using the refresh method

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/audit/ActionedByDto.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+import io.swagger.v3.oas.annotations.media.Schema
+import jakarta.validation.constraints.NotNull
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.enums.UserType
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.ActionedBy
+
+@Schema(description = "Actioned By")
+class ActionedByDto(
+
+  @Schema(description = "booker reference", example = "asd-aed-vhj", required = true)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/audit/ActionedByDto.kt:11]: required - false?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/TelemetryClientService.kt [hunk 10]
```diff
@@ -151,7 +151,9 @@ class TelemetryClientService(
 
     bookingEventAudit?.let {
       flagEventDataMap["visitBooked"] = formatDateTimeToString(it.createTimestamp)
-      flagEventDataMap["actionedBy"] = it.actionedBy
+      flagEventDataMap["actionedBy"] = bookingEventAudit.actionedBy.userName ?: ""
+      flagEventDataMap["bookerReference"] = bookingEventAudit.actionedBy.bookerReference ?: ""
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/TelemetryClientService.kt:155]: we could not send these values to appinsights if not null, but nothing defined on the story as such

**Review by dhirajshettymoj — COMMENTED** [review]
2 minor comments

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt [hunk 11]
```diff
@@ -1,58 +1,72 @@
...
 
   @Column(nullable = false)
   @Enumerated(EnumType.STRING)
-  var applicationMethodType: ApplicationMethodType,
+  val applicationMethodType: ApplicationMethodType,
 
-  @Column(nullable = false)
-  var actionedBy: String,
+  @Column(name = "ACTIONED_BY_ID", nullable = true)
+  private val actionedById: Long? = null,
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt:49]: If the actionedBy is a SYSTEM_USER, would it not have an actionedById?

Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt:49]: Basically, does this need to be nullable?

File: src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql (hunk 1 above)
Comment (adaviesMOJ) [reply src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql:15]: If this goes in before public booking goes live, it should be fine?

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/helper/EntityDataLoader.kt [hunk 12]
```diff
@@ -409,11 +414,12 @@ class VisitEntityHelper(
 @Transactional
 class EventAuditEntityHelper(
   private val eventAuditRepository: TestEventAuditRepository,
+  private val testActionedByRepository: TestActionedByRepository,
```
Comment (adaviesMOJ) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/helper/EntityDataLoader.kt:417]: Why do we have Test repositories? Why don't we use the actual repositories like we do in other Integration tests?

File: src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql (hunk 1 above)
Comment (AledWEvans) [reply src/main/resources/db/migration/V3_26__add_new_actioned_by_table.sql:15]: should go live soon,...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt (hunk 11 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt:49]: needs to be nullable = false,.... I think?

Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/EventAudit.kt:49]: Yes I think so.
PUBLIC / STAFF / SYSTEM, all of these will contain an Actioned_by_id, with different columns within the ActionedBy table being null...

But the ID field is mandatory right?

**Review by adaviesMOJ — APPROVED** [review]
